kolibri/utils/file_transfer.py

```diff
@@ -75,113 +84,390 @@ def retry_import(e):
+            current_chunk_file_size = (
+                os.path.getsize(chunk_file) if os.path.exists(chunk_file) else 0
+            )
+
+            with open(chunk_file, "ab") as f:
+                chunk_position = self.position % self.chunk_size
+                amount_to_write = min(remaining, self.chunk_size - chunk_position)
+                if chunk_position < current_chunk_file_size:
+                    diff = current_chunk_file_size - chunk_position
+                    chunk_position += diff
+                    self.position += diff
+                    amount_to_write -= diff
+                    remaining -= diff
+                f.seek(chunk_position)
+                to_write = data[
+                    len(data) - remaining : len(data) - remaining + amount_to_write
+                ]
+                # For some reason in Python 2.7 this is failing to return the number of bytes written.
+                # as we know exactly how much we are writing, we can just use that value.
+                f.write(to_write)
+                bytes_written = len(to_write)
+                self.position += bytes_written
+                remaining -= bytes_written
+
+    def read_data_until(self, end):
+        """
+        Generator to read data from the current position of the file until
+        but not including the end value.
+        Need to update this to give it an independent read position, as when this is called
+        the data is then written to the chunked file object (but ignored) so the position gets moved
+        twice leading to an overflow.
+        """
+        position = self.position
+        while position < end:
+            position, output = self._read(position, min(end - position, BLOCK_SIZE))
+            yield output
+
+    def _chunk_range_for_byte_range(self, start, end):
+        start_chunk = start // self.chunk_size if start is not None else 0
+        end_chunk = end // self.chunk_size if end is not None else self.chunks_count - 1
+        return start_chunk, end_chunk
+
+    def next_missing_chunk_and_read(self, start=None, end=None):
+        """
+        Generator to yield start and end ranges of the next missing chunk,
+        and return a generator to read the intervening data.
+        Exhausting the generator has no effect on the file position, as it is anticipated
+        that the data will be written back to the file in normal operation.
+        The data written back to the file will be ignored, but then the file position will be
+        updated as a result of the write.
+        """
+        start_chunk, end_chunk = self._chunk_range_for_byte_range(start, end)
+
+        self.seek(start_chunk * self.chunk_size)
+
+        with Cache(self.chunk_dir) as cache:
+
+            for chunk_index in range(start_chunk, end_chunk + 1):
+                chunk_file = self._get_chunk_file_name(chunk_index)
+                with Lock(cache, chunk_file):
```

Comment:
Should probably add an expiry here to prevent a permanent lock when trying to write to a specific file - maybe 60s?

Reply:
Yeah that's a good call.